File: observalog-node/src/__tests__/fingerprint.test.ts
import { computeFingerprint } from '../fingerprint';

describe('computeFingerprint', () => {
    it('is deterministic for same inputs', () => {
        const a = computeFingerprint(2, 'doc.storage.saved', 'IO_ERROR', 'doc123');
        const b = computeFingerprint(2, 'doc.storage.saved', 'IO_ERROR', 'doc123');
        expect(a).toBe(b);
    });

    it('ctx_primary_key prevents collision', () => {
        const a = computeFingerprint(2, 'doc.storage.saved', 'IO_ERROR', 'doc123');
        const b = computeFingerprint(2, 'doc.storage.saved', 'IO_ERROR', 'doc456');
        expect(a).not.toBe(b);
    });

    it('differs across service codes', () => {
        const a = computeFingerprint(1, 'auth.jwt.expired', 'TOKEN_EXPIRED', '');
        const b = computeFingerprint(2, 'auth.jwt.expired', 'TOKEN_EXPIRED', '');
        expect(a).not.toBe(b);
    });

    it('service code is raw byte, not ASCII digit', () => {
        // service code 1 (raw 0x01) must differ from ASCII '1' (0x31)
        const rawByte   = computeFingerprint(1,    'e', 'c', '');
        const asciChar  = computeFingerprint(0x31, 'e', 'c', '');
        expect(rawByte).not.toBe(asciChar);
    });

    it('returns a bigint', () => {
        const result = computeFingerprint(2, 'doc.storage.saved', 'IO_ERROR', '');
        expect(typeof result).toBe('bigint');
    });
});

File: observalog-java/src/test/java/com/observalog/FingerprintTest.java
package com.observalog;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class FingerprintTest {

    @Test
    void same_inputs_produce_same_hash() {
        long a = Fingerprint.compute((byte) 2, "doc.storage.saved", "IO_ERROR", "doc123");
        long b = Fingerprint.compute((byte) 2, "doc.storage.saved", "IO_ERROR", "doc123");
        assertEquals(a, b);
    }

    @Test
    void different_ctx_primary_key_prevents_collision() {
        long a = Fingerprint.compute((byte) 2, "doc.storage.saved", "IO_ERROR", "doc123");
        long b = Fingerprint.compute((byte) 2, "doc.storage.saved", "IO_ERROR", "doc456");
        assertNotEquals(a, b, "ctx_primary_key must disambiguate fingerprints");
    }

    @Test
    void different_service_codes_differ() {
        long a = Fingerprint.compute((byte) 1, "auth.jwt.expired", "TOKEN_EXPIRED", "");
        long b = Fingerprint.compute((byte) 2, "auth.jwt.expired", "TOKEN_EXPIRED", "");
        assertNotEquals(a, b);
    }

    @Test
    void service_code_is_raw_byte_not_ascii_digit() {
        // Service code 1 as raw byte (0x01) differs from ASCII '1' (0x31)
        long rawByte   = Fingerprint.compute((byte) 1,  "e", "c", "");
        long asciiChar = Fingerprint.compute((byte) '1', "e", "c", "");
        assertNotEquals(rawByte, asciiChar,
            "service code must be raw byte matching Go's string(uint8(code))");
    }
}
